### PR TITLE
MATModel: Model API + Tests for the Molecular Attention Transformer

### PR DESCRIPTION
**Pull request by atreyamaj**
# Pull Request Template

## Description

Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by rbharath** [comment]
`yapf` is failing and I'm seeing import errors on the PR. Could you take a look and make sure tests are passing?

File: deepchem/feat/molecule_featurizers/mat_featurizer.py [hunk 1]
```diff
@@ -2,6 +2,29 @@
 from deepchem.utils.molecule_feature_utils import one_hot_encode
 from deepchem.utils.typing import RDKitMol, RDKitAtom
 import numpy as np
+from typing import Tuple, Any
+from dataclasses import dataclass
+
+
+@dataclass
```
Comment (rbharath) [inline deepchem/feat/molecule_featurizers/mat_featurizer.py:9]: These changes are from the other MATFeaturizer PR right? Can you remove the changes from this PR to make it easier for me to review?

File: deepchem/models/torch_models/layers.py [hunk 2]
```diff
@@ -386,21 +407,24 @@ def forward(self, x: torch.Tensor, output: torch.Tensor) -> torch.Tensor:
     output: torch.Tensor
       Layer whose normalized output will be added to x.
     """
+    # if x is None:
```
Comment (rbharath) [inline deepchem/models/torch_models/layers.py:410]: Should this code be removed?

File: deepchem/models/torch_models/mat.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+               encoder_dropout_p: float = 0.0,
+               embed_input_hsize: int = 36,
+               embed_dropout_p: float = 0.0,
+               gen_aggregation_type: str = 'mean',
+               gen_dropout_p: float = 0.0,
+               gen_n_layers: int = 1,
+               gen_attn_hidden: int = 128,
+               gen_attn_out: int = 4,
+               gen_d_output: int = 1,
+               **kwargs):
```
Comment (rbharath) [inline deepchem/models/torch_models/mat.py:45]: Can you add a docstring for this class?

File: deepchem/models/torch_models/mat.py [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+from typing import Any
+
+
+class MAT(nn.Module):
+  '''An internal TorchModel class.
+
+  In this class, we define the various layers and establish a sequential model for the Molecular Attention Transformer.
+  We also define the forward call of this model in the forward function.
+
+  All arguments have the same meaning as in MATModel.
```
Comment (rbharath) [inline deepchem/models/torch_models/mat.py:16]: Can you add a usage example?

File: docs/source/atom_init.json [hunk 5]
```diff
@@ -0,0 +1 @@
...0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0], "97": [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], "98": [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], "99": [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], "100": [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]}
```
Comment (rbharath) [inline docs/source/atom_init.json:1]: This file should be removed

**Review by rbharath — COMMENTED** [review]
Did a first pass. Will do a more thorough pass once `MATFeaturizer` is merged in

File: docs/source/atom_init.json (hunk 5 above)
Comment (atreyamaj) [reply docs/source/atom_init.json:1]: I've removed this now.

File: deepchem/feat/molecule_featurizers/mat_featurizer.py (hunk 1 above)
Comment (atreyamaj) [reply deepchem/feat/molecule_featurizers/mat_featurizer.py:9]: I have removed changes in this and `test_mat_featurizer.py` as well from this PR now.

File: deepchem/models/torch_models/layers.py (hunk 2 above)
Comment (atreyamaj) [reply deepchem/models/torch_models/layers.py:410]: Sorry, I've removed this now.

File: deepchem/models/torch_models/mat.py [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+  We also define the forward call of this model in the forward function.
+
+  References
+  ----------
+  .. [1] Lukasz Maziarka et al. "Molecule Attention Transformer" Graph Representation Learning workshop and Machine Learning and the Physical Sciences workshop at NeurIPS 2019. 2020. https://arxiv.org/abs/2002.08264
+  
+  Examples
+  --------
+  >>> import deepchem as dc
+  >>> task, df, trans = dc.molnet.load_freesolv()
```
Comment (rbharath) [inline deepchem/models/torch_models/mat.py:23]: We shouldn't use molnet. We should use a simple set of smiles strings

File: deepchem/models/torch_models/mat.py [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+  Currently, MATModel is intended to be a regression model for the freesolv dataset.
+
+  References
+  ----------
+  .. [1] Lukasz Maziarka et al. "Molecule Attention Transformer" Graph Representation Learning workshop and Machine Learning and the Physical Sciences workshop at NeurIPS 2019. 2020. https://arxiv.org/abs/2002.08264
+
+  Examples
+  --------
+  >>> import deepchem as dc
+  >>> task, df, trans = dc.molnet.load_freesolv()
```
Comment (rbharath) [inline deepchem/models/torch_models/mat.py:189]: Same comment here

**Comment by atreyamaj** [comment]
Will put up a fresh PR, closing this for now.

**Pull request closed without merging**